### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by aloksGithub**
This repository contains the smart contract code to pull epoch rewards from stryke gauges and handle distribution of token rewards to Orange vault depositors using merkle trees

## Motivation

Stryke will be emitting SYK token rewards weekly via their gauges. Since Orange Finance invests its user's funds in Stryke, Orange would like to redirect the SYK rewards to depositors. Additionally, Orange Finance plans to distribute token rewards (ARB, USDC, etc.) to vault depositors in the future. Hence, a mechanism is required to be able to distribute the token rewards to Orange users based on their deposits and deposit durations.

## Solution

The Orange token distribution system will consist of
- Backend that calculates time weighted shares for Orange vault depositors and triggers reward pulling from gauges for SYK emissions. The time weighted shares and the total reward amount will be used to calculate the merkle root and proofs for reward claiming
- Smart contract that holds the token rewards and is updated with the merkle root. Users can provide a proof, to be able to claim their share of the token reward

### Roles and permissions

- [ ] It's Permissionless!
- [X] Roles and permissions listed below:
- Owner: Updates roots, sets vault gauge, triggers SYK pulling

### External Contracts

- [ ] No external contracts.
- [X] External contracts listed below:
- Stryke gauges:
    - 0x5f6D5a7e8eccA2A53C6322a96e9a48907A8284e0
    - 0x22dd31a495CafB229131A16C54a8e5b2f43C1162
    - 0xE32132282D181967960928b77236B3c472d5f396
- Stryke controller:
    - 0xFdf1B2c4E291b17f8E998e89cF28985fAF3cE6A1
- Orange vaults:
    - 0x5f6D5a7e8eccA2A53C6322a96e9a48907A8284e0
    - 0x22dd31a495CafB229131A16C54a8e5b2f43C1162
    - 0xE32132282D181967960928b77236B3c472d5f396"

### Deployment

Is this a new deployment or an upgrade to a deployed upgradable contract?

- [X] New deployment.
- [ ] Upgrade to a deployed upgradable contract:

### FUNDS AND PRIVILEGES

Does the contract hold any funds or privileges?

- [ ] No funds, allowance, or privileges.
- [X] Some:

## Checklist

- [X] I have performed a self-review of my code
- [X] If it is a new feature, I have added thorough tests.
- [ ] I have run your code through basic code analysis tools (Slither)
- [X] Solidity does not emit ANY compiler warnings.
- [X] The code is well-documented
- [X] The code has been deployed and tested on a forking network

### CODE REVIEW

File: contracts/OrangeDistributor.sol [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+     * @param _vault The vault for which to claim rewards
+     * @param _token The token to claim
+     * @param _amount Amount of token to claim as reward
+     * @param merkleProof Merkle proof
+     */
+    function claim(address _vault, address _token, uint256 _amount, bytes32[] calldata merkleProof) public {
+        claimed[_vault][_token][msg.sender]+=_amount;
+        
+        bytes32 leaf = keccak256(abi.encodePacked(msg.sender, claimed[_vault][_token][msg.sender]));
+        if (!MerkleProof.verifyCalldata(merkleProof, merkleRoot[_vault][_token], leaf)) revert InvalidProof();
```
Comment (maguroid) [inline contracts/OrangeDistributor.sol]: Just a heads up: You can use the require statement with custom errors starting from Solidity 0.8.27.
https://soliditylang.org/blog/2025/09/04/solidity-0.8.27-release-announcement/

File: contracts/OrangeDistributor.sol [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+        if (_token==syk) {
+            IERC20(syk).safeIncreaseAllowance(xSyk, _amount/2);
```
Comment (maguroid) [inline contracts/OrangeDistributor.sol:59]: Does the rounded value stay in the contract if the amount is odd?

File: contracts/TestContracts/GaugeType1.sol [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        // _disableInitializers();
+    }
+
+    /// @notice Initializes the gauge with necessary contract references and owner.
+    /// @param _initialOwner Address to be set as the owner of the contract.
+    /// @param _gaugeController Address of the GaugeController contract.
+    /// @param _syk Address of the SYK token contract.
+    function initialize(address _initialOwner, address _gaugeController, address _syk) public initializer {
```
Comment (maguroid) [inline contracts/TestContracts/GaugeType1.sol:38]: Unused argument is leading to compiler warnings.

File: contracts/OrangeDistributor.sol [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+     * @notice Claim all the token rewards for a vault
+     * @param _vault The vault for which to claim rewards
+     * @param _token The token to claim
+     * @param _amount Amount of token to claim as reward
+     * @param merkleProof Merkle proof
+     */
+    function claim(address _vault, address _token, uint256 _amount, bytes32[] calldata merkleProof) public {
+        claimed[_vault][_token][msg.sender]+=_amount;
+        
+        bytes32 leaf = keccak256(abi.encodePacked(msg.sender, claimed[_vault][_token][msg.sender]));
```
Comment (maguroid) [inline contracts/OrangeDistributor.sol:55]: Does the `claim` function work properly if a user doesn't claim for a specific epoch and then tries to claim in the next one?

File: contracts/SYKPuller.sol [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+     * @notice Pull the rewards for the next epoch from the gauge
+     */
+    function pullNext(address _vault) external onlyOwner {
+        uint balanceBefore = ERC20(syk).balanceOf(address(this));
+        IGauge(gauges[_vault]).pull(nextStrykeEpochToPull[_vault]);
+        uint balanceAfter = ERC20(syk).balanceOf(address(this));
+        epochRewards[_vault][nextStrykeEpochToPull[_vault]] = balanceAfter - balanceBefore;
+        nextStrykeEpochToPull[_vault]+=1;
+        emit RewardPulled(_vault, balanceAfter - balanceBefore);
+    }
```
Comment (maguroid) [inline contracts/SYKPuller.sol]: You can save gas by storing the epoch in memory.

```suggestion
    function pullNext(address _vault) external onlyOwner {
        uint nextEpoch = nextStrykeEpochToPull[_vault]++;
        uint balanceBefore = ERC20(syk).balanceOf(address(this));
        IGauge(gauges[_vault]).pull(nextEpoch);
        uint balanceAfter = ERC20(syk).balanceOf(address(this));
        epochRewards[_vault][nextEpoch] = balanceAfter - balanceBefore;
        emit RewardPulled(_vault, balanceAfter - balanceBefore);
    }
```

File: contracts/TestContracts/GaugeType1.sol (hunk 3 above)
Comment (aloksGithub) [reply contracts/TestContracts/GaugeType1.sol:38]: That contract is not part of the deployment, I am replacing some on chain code during testing. Please ignore all files in the TestContracts folder

File: contracts/OrangeDistributor.sol (hunk 4 above)
Comment (aloksGithub) [reply contracts/OrangeDistributor.sol:55]: Yes, if a user doesn't claim for an epoch, user can set amount to include multiple epochs

**Review by maguroid — APPROVED** [review]

**Pull request merged**